include/parallel_initialisation.hpp

```diff
@@ -0,0 +1,214 @@
+  particle_group->add_particle_dat(ParticleDat(
+      sycl_target, ParticleProp(Sym<REAL>("NESO_ORIG_POS"), space_ncomp),
+      domain->mesh->get_cell_count()));
+
+  double local_point[3];
+  get_point_in_local_domain(particle_group, local_point);
+
+  auto k_P = particle_group->position_dat->cell_dat.device_ptr();
+  auto k_ORIG_POS =
+      (*particle_group)[Sym<REAL>("NESO_ORIG_POS")]->cell_dat.device_ptr();
+  auto pl_iter_range =
+      particle_group->mpi_rank_dat->get_particle_loop_iter_range();
+  auto pl_stride =
+      particle_group->mpi_rank_dat->get_particle_loop_cell_stride();
+  auto pl_npart_cell =
+      particle_group->mpi_rank_dat->get_particle_loop_npart_cell();
+
+  sycl::buffer<double, 1> b_local_point(local_point, 3);
+
+  // store the target position on the particle and set the starting point.
+  sycl_target->queue
+      .submit([&](sycl::handler &cgh) {
+        auto a_local_point =
+            b_local_point.get_access<sycl::access::mode::read>(cgh);
+        cgh.parallel_for<>(sycl::range<1>(pl_iter_range), [=](sycl::id<1> idx) {
+          NESO_PARTICLES_KERNEL_START
+          const INT cellx = NESO_PARTICLES_KERNEL_CELL;
+          const INT layerx = NESO_PARTICLES_KERNEL_LAYER;
+          for (int nx = 0; nx < space_ncomp; nx++) {
+            k_ORIG_POS[cellx][nx][layerx] = k_P[cellx][nx][layerx];
+            k_P[cellx][nx][layerx] = a_local_point[nx];
+          }
+
+          NESO_PARTICLES_KERNEL_END
+        });
+      })
+      .wait_and_throw();
+}
+
+/**
+ *  Function used by parallel_advection_initialisation to step particles along a
+ * line to destination positions. Sets the particle positions to the original
+ * positions specified by the user.
+ *
+ *  @param particle_group ParticleGroup being initialised.
+ */
+inline void parallel_advection_restore(ParticleGroupSharedPtr particle_group) {
+
+  auto sycl_target = particle_group->sycl_target;
+  const int space_ncomp = particle_group->position_dat->ncomp;
+  auto k_P = particle_group->position_dat->cell_dat.device_ptr();
+  auto k_ORIG_POS =
+      (*particle_group)[Sym<REAL>("NESO_ORIG_POS")]->cell_dat.device_ptr();
+  auto pl_iter_range =
+      particle_group->mpi_rank_dat->get_particle_loop_iter_range();
+  auto pl_stride =
+      particle_group->mpi_rank_dat->get_particle_loop_cell_stride();
+  auto pl_npart_cell =
+      particle_group->mpi_rank_dat->get_particle_loop_npart_cell();
+
+  ErrorPropagate ep(sycl_target);
+  auto k_ep = ep.device_ptr();
+
+  // restore the target position
+  sycl_target->queue
+      .submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(sycl::range<1>(pl_iter_range), [=](sycl::id<1> idx) {
+          NESO_PARTICLES_KERNEL_START
+          const INT cellx = NESO_PARTICLES_KERNEL_CELL;
+          const INT layerx = NESO_PARTICLES_KERNEL_LAYER;
+
+          for (int nx = 0; nx < space_ncomp; nx++) {
+
+            const bool valid = ABS(k_ORIG_POS[cellx][nx][layerx] -
```

Comment:
Will we ever need both relative and absolute tolerance here?

Reply:
Updated to use absolute and relative error.